### PR TITLE
remove redundant imports

### PR DESCRIPTION
**Pull request by surechen**
detects redundant imports that can be eliminated.

for #117772 :

In order to facilitate review and modification, split the checking code and removing redundant imports code into two PR.

r? @petrochenkov

### CODE REVIEW

**Comment by petrochenkov** [comment]
r=me after removing `allow`s.
@rustbot author

**Comment by surechen** [comment]
> r=me after removing `allow`s. @rustbot author

Ok. Thank you.

File: compiler/rustc_data_structures/src/owned_slice.rs [hunk 1]
```diff
@@ -4,7 +4,12 @@ use crate::sync::Lrc;
 // Use our fake Send/Sync traits when on not parallel compiler,
 // so that `OwnedSlice` only implements/requires Send/Sync
 // for parallel compiler builds.
-use crate::sync::{Send, Sync};
+cfg_match! {
+    cfg(not(parallel_compiler)) => {
+        use crate::sync::{Send, Sync};
+    }
+    _ => {}
+}
```
Comment (petrochenkov) [inline compiler/rustc_data_structures/src/owned_slice.rs:12]: Unnecessary change.

File: library/core/src/intrinsics.rs [hunk 2]
```diff
@@ -2547,6 +2547,7 @@ macro_rules! assert_unsafe_precondition {
         }
     };
 }
+
```
Comment (petrochenkov) [inline library/core/src/intrinsics.rs:2550]: Unnecessary change.

**Comment by klensy** [comment]
This is only cleanup, actual implementation in other PR, why `relnotes` label?

**Comment by petrochenkov** [comment]
Yes, relnotes should be on https://github.com/rust-lang/rust/pull/117772.

**Comment by calebcartwright** [comment]
Following comment not based exclusively on this PR, but I'd like to again stress that non-mandatory changes to subtrees are strongly preferred to be made in the respective subtree repos by many of the subtree maintainers.

I feel like it's becoming far to common for subtrees to be unnecessarily modified in this repo for the sake of convenience and expediency, and we then often end up paying for that expediency on the subtree side during syncs.

**Comment by matthiaskrgr** [comment]
@bors r- rollup=iffy 
failed in a rollup https://github.com/rust-lang/rust/pull/118757#issuecomment-1847904478

File: compiler/rustc_data_structures/src/owned_slice.rs [hunk 3]
```diff
@@ -4,7 +4,6 @@ use crate::sync::Lrc;
 // Use our fake Send/Sync traits when on not parallel compiler,
 // so that `OwnedSlice` only implements/requires Send/Sync
 // for parallel compiler builds.
-use crate::sync::{Send, Sync};
```
Comment (petrochenkov) [inline compiler/rustc_data_structures/src/owned_slice.rs:7]: Needs to be reverted (https://github.com/rust-lang/rust/pull/118692#issuecomment-1847908699).

**Comment by petrochenkov** [comment]
@calebcartwright 
I've seen this stated before, but I don't really agree with it with regards to changes like these.
Rebasing rustfmt subtree over changes like this during update is trivial, cherry-picking these changes to rusfmt repo would also be  trivial, but doing an additional rustfmt update and landing it into rust-lang/rust to unblock https://github.com/rust-lang/rust/pull/117772 is not trivial.

Allowing `unused_imports` in the whole rustfmt to minimize changes also doesn't seem right, given that the proper solution (removing a few imports) in not any more complex.

File: compiler/rustc_data_structures/src/owned_slice.rs (hunk 3 above)
Comment (surechen) [reply compiler/rustc_data_structures/src/owned_slice.rs:7]: Thank you. Because of the direct revert, the unused imports check will report an error. Is it possible to distinguish this situation through the cfg macro?

Comment (petrochenkov) [reply compiler/rustc_data_structures/src/owned_slice.rs:7]: What I wanted to do in this PR is to apply trivial changes only (unconditional import removals).
False positives, allows, conditional imports can all be deferred to https://github.com/rust-lang/rust/pull/117772.
That's why I suggested reverting this removal in the comment above, and not adding a cfg.

Comment (petrochenkov) [reply compiler/rustc_data_structures/src/owned_slice.rs:7]: In this specific case explicit `sync::Send` should be used instead of an import, that's how it's done in all other places in the compiler.

Comment (surechen) [reply compiler/rustc_data_structures/src/owned_slice.rs]: > In this specific case explicit `sync::Send` should be used instead of an import, that's how it's done in all other places in the compiler.

Got it, thank you for your patient explanation.

**Comment by calebcartwright** [comment]
@petrochenkov 

re: https://github.com/rust-lang/rust/pull/118692#issuecomment-1847921800

Thanks, I genuinely appreciate you not only taking the time to reply, but engaging in discussion of the topic! :heart:  :pray: 

You're correct that I've expressed this sentiment before, and I've also seen it echoed/shared by other subtree maintainers elsewhere. I confess I've found it confusing and occasionally frustrating when those sentiments are outright ignored or quickly dismissed.

This is probably a topic best to continue discussing in Zulip, but suffice to say that I feel the dev guide has [prescribed guidance on this topic](https://rustc-dev-guide.rust-lang.org/external-repos.html?highlight=subtree#external-dependencies-subtree) which is often contradicted in practice, and at a minimum I think that's something which should be addressed (even if that means reversing or clarifying the language in the dev guide)

> Rebasing rustfmt subtree over changes like this during update is trivial,

fwiw this is incorrect, rebasing cannot be used on the subtree pushes which is a big part of why i keep raising this issue. it does produce very [real challenges](https://rust-lang.zulipchat.com/#narrow/stream/357797-t-rustfmt/topic/meeting.202023-10-02/near/395999315) for us, i promise i'm not just making noise on the basis of speculation :sweat_smile:

**Comment by lnicola** [comment]
On the rust-analyzer side, I'm very worried about merging back these changes because we've had problems with `git-subtree` in the past, and it seems a little janky at times (it's been saying for about 6 months that it's resolving one specific conflict on every sync back). In practice, though, merging back trivial PRs like this one hasn't been an issue yet.

It does feel, however, like they're becoming more and more common lately, and it's sometimes even hard to keep with them. I understand build fixes, but doing this for cosmetic changes really feels unnecessary. Lints in particular are a bit weird -- on the RA side we're fine with the stable ones. If it's that important for rust-lang/rust to run unstable lints against subtrees, okay, we'll handle it, but it's pretty stressful.

PS: I noticed recently that syncing back yields a ton of conflicts if the upstream (RA) has new commits. But if you do it in two steps (merge up to the version that's in downstream, then merge again with `master`), it's much better. Just something to keep in mind.

**Comment by calebcartwright** [comment]
Well said @lnicola

For me it's not about any one particular change in isolation, it's very much the aggregate set of changes that I find often has a compounding effect on complicating the subtree syncs and the increase in frequency of such changes. git subtree is indeed very touch and, for whatever reason, quickly and easily gets very fussy
